table_engine/src/partition/rule/key.rs

```diff
@@ -0,0 +1,346 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+//! Key partition rule
+
+use std::collections::{HashMap, HashSet};
+
+use common_types::{
+    bytes::{BufMut, BytesMut},
+    datum::Datum,
+    hash::hash64,
+    row::{Row, RowGroup},
+};
+use itertools::Itertools;
+use log::debug;
+use snafu::OptionExt;
+
+use crate::partition::{
+    rule::{filter::PartitionCondition, ColumnWithType, PartitionFilter, PartitionRule},
+    LocateReadPartition, LocateWritePartition, Result,
+};
+
+pub struct KeyRule {
+    pub typed_columns: Vec<ColumnWithType>,
+    pub partition_num: u64,
+}
+
+impl KeyRule {
+    fn get_candidate_partition_keys_groups(
+        &self,
+        filters: &[PartitionFilter],
+    ) -> Result<Vec<Vec<usize>>> {
+        let column_name_to_idxs = self
+            .typed_columns
+            .iter()
+            .enumerate()
+            .map(|(col_idx, typed_col)| (typed_col.column.clone(), col_idx))
+            .collect::<HashMap<_, _>>();
+        let mut filter_by_columns = vec![Vec::new(); self.typed_columns.len()];
+
+        // Group the filters by their columns.
+        for (filter_idx, filter) in filters.iter().enumerate() {
+            let col_idx = column_name_to_idxs
+                .get(filter.column.as_str())
+                .with_context(|| LocateReadPartition {
+                    msg: format!(
```

Comment:
Don't return error, ignore the filter is ok?

Reply:
I still keep returning error now, but make it an internal error.
It will be catch in `locate_partitions_for_read`, and then report it by log and return the all partitions.